clientv3/watch.go

```diff
@@ -41,3 +46,480 @@ type WatchResponse struct {
+	w.mu.RLock()
+	defer w.mu.RUnlock()
+	ws, ok := w.streams[pbresp.WatchId]
+	if ok {
+		wr := &WatchResponse{*pbresp.Header, pbresp.Events}
+		ws.recvc <- wr
+	}
+	return ok
+}
+
+// serveWatchClient forwards messages from the grpc stream to run()
+func (w *watcher) serveWatchClient(wc pb.Watch_WatchClient) {
+	for {
+		resp, err := wc.Recv()
+		if err != nil {
+			select {
+			case w.errc <- err:
+			case <-w.donec:
+			}
+			return
+		}
+		select {
+		case w.respc <- resp:
+		case <-w.donec:
+			return
+		}
+	}
+}
+
+// serveCtxs removes a watcher when its watch context cancels
+func (w *watcher) serveCtxs() {
+	for {
+		cases := []reflect.SelectCase{}
+		wss := []*watcherStream{}
+
+		// build list of ctx done channels
+		w.mu.RLock()
+		sc := reflect.SelectCase{Dir: reflect.SelectRecv}
+		for _, ws := range w.streams {
+			wss = append(wss, ws)
+			sc.Chan = reflect.ValueOf(ws.initReq.ctx.Done())
+			cases = append(cases, sc)
+		}
+		w.mu.RUnlock()
+
+		// add control channels
+		sc.Chan = reflect.ValueOf(w.ctxc)
+		cases = append(cases, sc)
+		sc.Chan = reflect.ValueOf(w.donec)
+		cases = append(cases, sc)
+
+		idx, _, _ := reflect.Select(cases)
+		switch {
+		// w.ctxc; refresh list
+		case idx == len(cases)-2:
+		// w.donec; shutdown
+		case idx == len(cases)-1:
+			return
+		// stream cancelled
+		default:
+			w.mu.Lock()
+			w.closeStream(wss[idx])
+			w.mu.Unlock()
+			// lazily send cancel message if events on missing id
+		}
+	}
+}
+
+// serveStream forwards watch responses from run() to the subscriber
+func (w *watcher) serveStream(ws *watcherStream) {
+	emptyWr := &WatchResponse{}
+	wrs := []*WatchResponse{}
+	for {
+		curWr := emptyWr
+		outc := ws.outc
+		if len(wrs) > 0 {
+			curWr = wrs[0]
+		} else {
+			outc = nil
+		}
+		select {
+		case outc <- *curWr:
+			newRev := wrs[0].Events[0].Kv.ModRevision
+			if newRev != ws.lastRev {
+				ws.revEvents = 0
+				ws.lastRev = newRev
+			}
+			ws.revEvents += len(wrs[0].Events)
+			wrs[0] = nil
+			wrs = wrs[1:]
+		case wr, ok := <-ws.recvc:
+			if !ok {
+				// shutdown from closeStream
+				return
+			}
+			wrs = append(wrs, wr)
+		case <-w.donec:
+			return
+		}
+	}
+}
+
+func (w *watcher) newWatchClient() (pb.Watch_WatchClient, error) {
+	ws, rerr := w.resume()
+	if rerr != nil {
+		return nil, rerr
+	}
+	go w.serveWatchClient(ws)
+	return ws, nil
+}
+
+// resume creates a new WatchClient with all current watchers resestablished
+func (w *watcher) resume() (ws pb.Watch_WatchClient, err error) {
+	for {
+		if ws, err = w.openWatchClient(); err != nil {
+			break
+		} else if err = w.resumeWatchers(ws); err == nil {
+			break
+		}
+	}
+	return ws, err
+}
+
+// openWatchClient retries opening a watchclient until retryConnection fails
+func (w *watcher) openWatchClient() (ws pb.Watch_WatchClient, err error) {
+	for {
+		if ws, err = w.remote.Watch(w.ctx); ws != nil {
+			break
+		} else if isRPCError(err) {
+			return nil, err
+		}
+		newConn, nerr := w.c.retryConnection(w.conn, nil)
```

Comment:
This will cause the watch API to be closed and all request will be canceled? the future use of the watchAPI will return an error? I am fine with this, but we probably would like to see a deterministic behavior. (also we need to make lease/watch behavior consistently)